src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -764,6 +865,17 @@ public void RemoveFromWorkspace()
         {
             _documentFileChangeContext.Dispose();
 
+            lock (_gate)
+            {
+                // clear tracking to external components
+                foreach (var provider in _eventSubscriptionTracker)
+                {
+                    provider.Updated -= OnDynamicFileInfoUpdated;
+                }
+
+                _eventSubscriptionTracker.Clear();
+            }
```

Comment:
why is this in the lock but the other code in RemoveFromWorkspace is not?

Reply:
@jasonmalinowski why other fields are not under lock? what gurantees order of events for those fields?